### PR TITLE
Adaptive Nelder-Mead

### PR DESCRIPTION
**Pull request by juliusbierk**
A small suggestive addition for the Nelder-Mead algortihm.

This paper [Sec 4.1] https://link.springer.com/article/10.1007%2Fs10589-010-9329-3 describes how adapting the step-parameters to the dimensionality of the problem at hand improves performance, especially for high-dimensional problems.
This pull request adds this as an option with the optional adaptive=True/False argument.

I don't know if this is something that is wanted in scipy, but I'm using this version myself, and so I thought I'd make a pull request.

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @juliusbierk. Could you please add a unit test for the new keyword?

**Comment by rgommers** [comment]
TravisCI build failure is unrelated

**Comment by rgommers** [comment]
Would be good to add the paper this is based on to the ``References`` section in the function docstring, and reference it in the keyword description.

**Comment by juliusbierk** [comment]
Sorry it's taken a while, @rgommers. I've added the reference and a unit test that only passes with the adaptive=True parameter.

File: scipy/optimize/_minimize.py [hunk 1]
```diff
@@ -307,7 +307,11 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
        Center -- Institute for Flight Mechanics, Koln, Germany.
     .. [13] Conn, A. R., Gould, N. I., and Toint, P. L.
        Trust region methods. 2000. Siam. pp. 169-200.
-
+    .. [14] Gao, F. and Han, L.
```
Comment (larsoner) [inline scipy/optimize/_minimize.py]: If you add a ref here it's best to reference it in the text above

**Review by larsoner — COMMENTED** [review]
@rgommers mentioned a keyword description -- you should add one to (at least) `fmin`

File: scipy/optimize/_minimize.py (hunk 1 above)
Comment (dlax) [reply scipy/optimize/_minimize.py]: I guess the reference should be in `_minimize_neldermead()` below where the citation is added.

**Comment by juliusbierk** [comment]
I hope I understood what you meant correctly. I've updated the PR accordingly.

**Comment by juliusbierk** [comment]
@larsoner Is this update what you meant?

File: scipy/optimize/tests/test_optimize.py [hunk 2]
```diff
@@ -443,6 +443,14 @@ def test_neldermead_xatol_fatol():
...
                  func, [1, 1], xtol=1e-3, ftol=1e-3, maxiter=2)
 
+def test_neldermead_adaptive():
+    func = lambda x: np.sum(x**2)
+    p0 = [0.15746215, 0.48087031, 0.44519198, 0.4223638, 0.61505159, 0.32308456,
+      0.9692297, 0.4471682, 0.77411992, 0.80441652, 0.35994957, 0.75487856,
+      0.99973421, 0.65063887, 0.09626474]
+    res = optimize.minimize(func, p0, method='Nelder-Mead',
+                    options={'adaptive':True})
+    assert_equal(res.success, True)  # res.success == False for non-adaptive
```
Comment (larsoner) [inline scipy/optimize/tests/test_optimize.py]: Might be worth explicitly doing the non-adaptive, too, and `assert_equal(res.succes, False)`?

**Comment by juliusbierk** [comment]
I thought about that, too. I wasn't sure, however, that you wanted a negative test like that on minimising, in case of future improvements. But I guess the underlying algorithm won't be changed...

**Comment by juliusbierk** [comment]
@larsoner I added the test.

File: scipy/optimize/optimize.py [hunk 3]
```diff
@@ -433,7 +434,17 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
     fatol : number, optional
         Absolute error in func(xopt) between iterations that is acceptable for
         convergence.
+    adaptive : bool, optional
+        Adapt algorithm parameters to dimensionality of problem. Useful for
+        high-dimensional minimization [1].
```
Comment (larsoner) [inline scipy/optimize/optimize.py]: I think this needs to be `[1]_` to link properly

**Review by larsoner — APPROVED** [review]
LGTM but it might be better if someone who works more closely with minimizers could have a look

File: scipy/optimize/optimize.py (hunk 3 above)
Comment (juliusbierk) [reply scipy/optimize/optimize.py]: Right, thanks!

**Comment by pv** [comment]
thanks, merged

**Pull request merged**